### PR TITLE
feat(eventing): add nexus events to stats module

### PR DESCRIPTION
**Pull request by datacore-vvarakantham**
Added nexus events to stats module.
Added nexus stats metrics to transmission module.

## Description

The nexus events added consists of creation, deletion counts, and rebuild begun and ended counts.
## How Has This Been Tested?

Manually tested.
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.

Dependency on https://github.com/openebs/mayastor-control-plane/pull/673

### CODE REVIEW

File: call-home/src/bin/stats/exporter/events_collector.rs [hunk 1]
```diff
@@ -108,6 +119,47 @@ impl StatsCollector {
         metric_family.extend(pools_deleted.collect().pop());
         metric_family
     }
+
+    fn nexus_metrics(&self, events: &EventSet) -> Vec<prometheus::proto::MetricFamily> {
+        let mut metric_family = Vec::new();
+        let nexus_created = match self.nexus.get_metric_with_label_values(&[CREATED]) {
+            Ok(nexus) => nexus,
+            Err(error) => {
+                error!(%error,"Error while creating metrics(nexus created) with label values: {}", CREATED );
```
Comment (tiagolobocastro) [inline call-home/src/bin/stats/exporter/events_collector.rs:128]:
```suggestion
                error!(%error,"Error while creating metrics(nexus created) with label values: {CREATED}");
```

File: call-home/src/bin/stats/exporter/events_collector.rs [hunk 2]
```diff
@@ -108,6 +119,47 @@ impl StatsCollector {
...
+        rebuild_begun.inc_by(events.nexus.rebuild_begun as f64);
+        let rebuild_ended = match self.nexus.get_metric_with_label_values(&[REBUILD_ENDED]) {
+            Ok(nexus) => nexus,
+            Err(error) => {
+                error!(%error,"Error while creating metrics(rebuild ended) with label values: {}", REBUILD_ENDED);
+                return metric_family;
+            }
+        };
+        rebuild_ended.inc_by(events.nexus.rebuild_ended as f64);
+        metric_family.extend(nexus_created.collect().pop());
```
Comment (tiagolobocastro) [inline call-home/src/bin/stats/exporter/events_collector.rs:157]: Seems like a funny syntax, why are we collecting an popping?

Comment (datacore-vvarakantham) [reply call-home/src/bin/stats/exporter/events_collector.rs:157]: The collect() here is collects the single metric as a vector. But I guess pop() is not required here. Will check.

Comment (datacore-vvarakantham) [reply call-home/src/bin/stats/exporter/events_collector.rs]: pop() is not need. Removed it.

File: call-home/src/bin/stats/exporter/events_collector.rs (hunk 1 above)
Comment (datacore-vvarakantham) [reply call-home/src/bin/stats/exporter/events_collector.rs]: Done

File: call-home/src/bin/stats/exporter/events_collector.rs (hunk 2 above)
Comment (tiagolobocastro) [reply call-home/src/bin/stats/exporter/events_collector.rs]: Before removing it, let's understand why it was there. Was this a 1-element vector?

Comment (datacore-vvarakantham) [reply call-home/src/bin/stats/exporter/events_collector.rs]: Yes. it is a single element vector.

File: call-home/src/bin/callhome/collector/report_models.rs [hunk 3]
```diff
@@ -302,6 +333,8 @@ pub(crate) enum Action {
     Unknown,
     Created,
     Deleted,
+    RebuildBegun,
```
Comment (sinhaashish) [inline call-home/src/bin/callhome/collector/report_models.rs:336]: IMO `RebuildBegun` -> `RebuildStart`

File: chart/templates/mayastor/io/io-engine-daemonset.yaml [hunk 4]
```diff
@@ -98,7 +98,8 @@ spec:
         - "-T={{ .Values.io_engine.target.nvmf.iface }}"{{ end }}{{ if .Values.io_engine.envcontext }}
         - "--env-context=--{{ .Values.io_engine.envcontext }}"{{ end }}{{ if .Values.io_engine.reactorFreezeDetection.enabled }}
         - "--reactor-freeze-detection"{{ end }}
-        - "--tgt-crdt={{ .Values.io_engine.target.nvmf.hostCmdRetryDelay.crdt1 }}"
+        - "--tgt-crdt={{ .Values.io_engine.target.nvmf.hostCmdRetryDelay.crdt1 }}"{{ if .Values.eventing.enabled }}
+        - "--events-url=nats://{{ .Release.Name }}-nats:4222"{{ end }}
```
Comment (sinhaashish) [inline chart/templates/mayastor/io/io-engine-daemonset.yaml:102]: Whys is this added ?

File: call-home/src/common/constants.rs [hunk 5]
```diff
@@ -40,12 +43,21 @@ pub const CREATED: &str = "created";
 /// Delete actions for events.
 pub const DELETED: &str = "deleted";
 
+/// rebuild begun action for events.
```
Comment (sinhaashish) [inline call-home/src/common/constants.rs:46]: Start comment with capital `rebuild` -> `Rebuild`

File: chart/templates/mayastor/io/io-engine-daemonset.yaml (hunk 4 above)
Comment (datacore-vvarakantham) [reply chart/templates/mayastor/io/io-engine-daemonset.yaml:102]: It is needed to enable eventing in io-engine and pass mbus-url

Comment (niladrih) [reply chart/templates/mayastor/io/io-engine-daemonset.yaml:102]: Should the eventing flags be inside an if like the ones on the eventing components? Not just this one, all of the others as well?
```suggestion
{{- if .Values.eventing.enabled }}
        - "--events-url=nats://{{ .Release.Name }}-nats:4222"{{ end }}
{{- end }}        
```

Comment (datacore-vvarakantham) [reply chart/templates/mayastor/io/io-engine-daemonset.yaml:102]: Only when the "events-url" is passed, the eventing is enabled in the core agent or io-engine. So I am passing "events-url" when eventing is enabled.

Comment (niladrih) [reply chart/templates/mayastor/io/io-engine-daemonset.yaml:102]: Oh, okay the if's are over to the right. Sorry, I didn't notice it. Great!

**Review by tiagolobocastro — APPROVED** [review]

**Review by sinhaashish — APPROVED** [review]

**Comment by datacore-vvarakantham** [comment]
bors merge

**Pull request merged**